### ISSUE TITLE
Script failed with error : 'JsErrorScriptException (0x30001)' during MSBuild

### ISSUE THREAD

**Issue by DerFlatulator** [Microsoft/TypeScript#9720]
**TypeScript Version:**  2.0.0

I'm running MSBuild From VS2015 Update 3 with a large TypeScript Project.

The following is in csproj (tried `2.0` too):

```xml
<TypeScriptToolsVersion>1.9</TypeScriptToolsVersion>
``` 

tsconfig:

```json
{
    "compilerOptions": {
        "noImplicitAny": false,
        "removeComments": false,
        "preserveConstEnums": true,
        "target": "es5",
        "sourceMap": false,
        "experimentalDecorators": true,
        "pretty": true,
        "noEmitOnError": false
    },
    "compileOnSave": true,
    "files": [
        // many files
    ]
}
```

**Expected behavior:**

Successful compilation. I can get it to compile on one machine, but not another.

**Actual behavior:** 

```
[16:16:28]Unknown output : Script failed with error : 'JsErrorScriptException (0x30001)'. [R:\BuildAgent\work\Core.csproj]
[16:16:28]Unknown output : error : Incorrect function. [R:\BuildAgent\work\Core.csproj]
```

Some more info:

```
Error: Incorrect function.

   at ChakraHost.realpath (Unknown script code:1538:70)
   at ts.sys.realpath (Unknown script code:36513:62)
   at nodeModuleNameResolver (Unknown script code:36261:13)
   at resolveModuleName (Unknown script code:36058:17)
   at loader_1 (Unknown script code:36630:68)
   at loadWithLocalCache (Unknown script code:36576:17)
   at resolveModuleNamesWorker (Unknown script code:36631:81)
   at processImportedModules (Unknown script code:37357:17)
   at findSourceFile (Unknown script code:37281:17)
   at processSourceFile (Unknown script code:37201:22)
```

Full Visual Studio Info:

```
Microsoft Visual Studio Community 2015
Version 14.0.25420.01 Update 3
Microsoft .NET Framework
Version 4.6.01055

Installed Version: Community

Visual Basic 2015   00322-20000-00000-AA159
Microsoft Visual Basic 2015

Visual C# 2015   00322-20000-00000-AA159
Microsoft Visual C# 2015

Visual C++ 2015   00322-20000-00000-AA159
Microsoft Visual C++ 2015

Application Insights Tools for Visual Studio Package   1.0
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2015.1 (Beta8)   14.1.11107.0
ASP.NET and Web Tools 2015.1 (Beta8)

ASP.NET Web Frameworks and Tools 2013   5.2.40314.0
For additional information, visit http://www.asp.net/

Common Azure Tools   1.8
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

JavaScript Language Service   2.0
JavaScript Language Service

JavaScript Project System   2.0
JavaScript Project System

Microsoft Azure Mobile Services Tools   1.4
Microsoft Azure Mobile Services Tools

NuGet Package Manager   3.4.4
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

PreEmptive Analytics Visualizer   1.2
Microsoft Visual Studio extension to visualize aggregated summaries from the PreEmptive Analytics product.

SQL Server Data Tools   14.0.50616.0
Microsoft SQL Server Data Tools

TypeScript   2.0.0.0
TypeScript tools for Visual Studio
```

**Comment by mhegazy** [comment]
Can you please install [Update 3](https://www.visualstudio.com/en-us/news/releasenotes/vs2015-update3-vs), reinstall TypeScript 2.0, and update the version in your project to `2.0` instead of `1.9`. and let us know if you are still seeing the issue.

**Comment by DerFlatulator** [comment]
Apologies, that was a typo in the original post. The above error was with Update **3**, TS2.0 and `TypeScriptToolsVersion` set to both 1.9 and 2.0.

On the machine(s) that are having this issue, MS Edge 20 is installed. This error does not occur on a machine with Edge 25. Could it be an outdated version of Chakra Core that is causing this issue?

**Comment by DerFlatulator** [comment]
Additional note, our workaround is to use npm's `typescript@2.0.0` (with v8) to compile the files that MSBuild cannot.

**Comment by mhegazy** [comment]
the error indicate a mismatch between the js file and the tsc.exe file. can you share the version of the `tsc.exe` file as well?

**Comment by DerFlatulator** [comment]
Both `tsc.exe` and `tsc.js` report `2.0.0`.

```
C:\>where tsc
C:\Program Files (x86)\Microsoft SDKs\TypeScript\2.0\tsc.exe
C:\Program Files (x86)\Microsoft SDKs\TypeScript\2.0\tsc.js
```

**Comment by NoelAbrahams** [comment]
I have a repro for this on my machine - not sure if it is universal.

* Create a new HTML Application with TypeScript in VS 2015 Update 3.
* Add the following code

 ```ts
declare const Symbol: SymbolConstructor;

interface IterableIterator<T> {

   [Symbol.iterator](): IterableIterator<T>;
}

interface SymbolConstructor {

    iterator: symbol;
}
```
* Compiling the project result in the following:

```
1>------ Rebuild All started: Project: TypeScriptHTMLApp3, Configuration: Debug Any CPU ------
1>Unknown output : Script failed with error : 'JsErrorScriptException (0x30001)'.
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
```

The offending bit of code is the reference to `Symbol.iterator`.

The workaround is to comment out (in lib.d.ts) all references in index signatures that contain `Symbol.`

Repro project attached.

[TypeScriptHTMLApp3.zip](https://github.com/Microsoft/TypeScript/files/391057/TypeScriptHTMLApp3.zip)

**Comment by engineforce** [comment]
The issue reported by DerFlatulator is different from NoelAbrahams, I am from the same team of DerFlatulator. The problem is not related to Msbuild or Visual Studio. 

On the problem machine: Windows Server 2012 R2 Standard, I was able to reproduce the same issue using tsc.exe command line tool.

```
// module1.ts
import module2 = require("./module2")

export module module1 {
	console.log("Module 1");	
}

// module2.ts
export module module2 {
	console.log("Module 2");
}
```

Error:

```
R:\TeamCity\Debug>"C:\Program Files (x86)\Microsoft SDKs\TypeScript\2.0\tsc.exe" module1.ts
Script failed with error: 'JsErrorScriptException (0x30001)'.
Incorrect function.

Error: Incorrect function.

   at ChakraHost.realpath (Unknown script code:1538:70)
   at ts.sys.realpath (Unknown script code:36513:62)
   at nodeModuleNameResolver (Unknown script code:36261:13)
   at resolveModuleName (Unknown script code:36058:17)
   at loader_1 (Unknown script code:36630:68)
   at loadWithLocalCache (Unknown script code:36576:17)
   at resolveModuleNamesWorker (Unknown script code:36631:81)
   at processImportedModules (Unknown script code:37357:17)
   at findSourceFile (Unknown script code:37281:17)
```

**Comment by DerFlatulator** [comment]
@engineforce is `R:\` above a RAM drive? It was is on the CI machines that reported this issue, and may be relevant since this looks like a module resolution or file system related problem.

**Comment by engineforce** [comment]
@DerFlatulator: you are correct, R: is a RAM disk, and the error only happens on this drive. If I moved module1.ts and module2.ts (from above comment) to D: drive, then no error occurred. 
@mhegazy: can this be fixed by TypeScript team? RAM disk has never caused a issue before 2.0.

**Comment by mhegazy** [comment]
We are not aware of any changes that impact RAM disks specifically. what version of Node are you using?

**Comment by mhegazy** [comment]
@zhengbli can you take a look

**Comment by josundt** [comment]
I experience the same problem when trying to compile my ASP.NET 4.6.1 project with TypeScript MSBUILD v.2.0.2-rc when the building on an Azure Web Site after deploying from GitHub.
Microsoft.TypeScript.MSBuild.2.0.2-rc.

I have investigated some, and I found that when I compile using the "tsc.exe" file in the NuGet packages folder under Microsoft.TypeScript.MSBuild.2.0.2-rc, it always fails when running the command in Kudu on the Azure Web Site (probably a Windows Server 2012 R2), but the excact same compiles fine on my Windows 10 Anniversary Update desktop.
Microsoft.TypeScript.MSBuild.2.0.2-rc.

The tsc.exe included with the MSBuild integration runs on top of the ChakraCore runtime instead of Node. I have tested compiling with the node tsc, and then it works both places.

I expect this to be a ChakraCore specific problem on Windows 2012 R2 Server that you will not experience on Windows 10.

**Comment by DerFlatulator** [comment]
@mhegazy Node.js installed on the affected machines varies from 4.X to 5.X.

@josundt These seem like similar, but slightly different issues. The first rises as an `Incorrect Function` error, the second is a `TypeError`. The original issue affects Windows 8 and 10.

**Comment by mxa0079** [comment]
I just observed this behavior in VS2015 Update 3 with TS 1.8.36.0

My workaround was to comment out "watch" : true in the compiler options:

```
{
  "compileOnSave": true,
  "compilerOptions": {
    "noImplicitAny": false,
    "removeComments": false,
    "target": "ES5"//,
    //"watch": true
  }
}
```

**Comment by josundt** [comment]
@DerFlatularor These may be different bugs, but they both seem to be chakracore specific and the errors occur while building the compiler source files list

**Comment by mihailik** [comment]
Could this be another manifestation of missing DLL problem?

https://github.com/Microsoft/TypeScript/issues/7335

**Comment by mhegazy** [comment]
this seems to be the same issue in https://github.com/Microsoft/TypeScript/issues/11110. we have traced this back to Chakra issue on Windows Server 2012, as well as other OS's. we are following up on the issue and should have an update shortly.

**Comment by mhegazy** [comment]
Turns out this is a different issue from #11110. So we still do not know what is the cause of this. We are still trying to reproduce this locally. Should have an update soon.

**Comment by yuit** [comment]
Thanks @vladima for helping me investigating the original reported issue from @DerFlatulator  and @engineforce. The issue is from when realPath function (used by node-module-resolution)  call "GetFinalPathNameByHandle" in trying to get finalPath in Dos naming. However, due to how RamDisk is created with ImDisk, "GetFinalPathNameByHandle" unsuccessfully find the path in Dos naming. @Vlad has  suggested a workaround in majority of the cases

**Comment by mhegazy** [comment]
The fix should be published in the TypeScript 2.0.6.

**Comment by mhegazy** [comment]
[TypeScript 2.0.6](https://github.com/Microsoft/TypeScript/releases/tag/v2.0.6) should be available now; please give it a try and let us know if there are other issues.
